eosmetrics/emtr-aggregate-timer.c

```diff
@@ -0,0 +1,163 @@
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * eos-metrics is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-metrics.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "emtr-aggregate-timer-private.h"
+
+
+/**
+ * SECTION:emtr-aggregate-timer
+ * @title: Aggregate events timer
+ * @short_description: Timer to aggregate events.
+ * @include: eosmetrics/eos-aggregate-timer.h
+ *
+ * EmtrAggregateTimer is a simple timer to aggregate events.
+ */
+
+struct _EmtrAggregateTimer
+{
+  GObject parent_instance;
+
+  GCancellable *cancellable; /* (owned) */
+  EmerAggregateTimer *timer_proxy; /* (owned) */
+  EmerEventRecorderServer *dbus_proxy; /* (unowned) */
+
+  gboolean stopped;
+};
+
+G_DEFINE_TYPE (EmtrAggregateTimer, emtr_aggregate_timer, G_TYPE_OBJECT)
+
+static void
+emtr_aggregate_timer_finalize (GObject *object)
+{
+  EmtrAggregateTimer *self = (EmtrAggregateTimer *)object;
+
+  g_cancellable_cancel (self->cancellable);
+  g_clear_object (&self->cancellable);
+
+  if (!self->stopped && self->timer_proxy)
+    {
+      g_autoptr (GError) error = NULL;
+      emer_aggregate_timer_call_stop_timer_sync (self->timer_proxy, NULL, &error);
+    }
+
+  g_clear_object (&self->timer_proxy);
+
+  G_OBJECT_CLASS (emtr_aggregate_timer_parent_class)->finalize (object);
+}
+
+static void
+emtr_aggregate_timer_class_init (EmtrAggregateTimerClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = emtr_aggregate_timer_finalize;
+}
+
+static void
+emtr_aggregate_timer_init (EmtrAggregateTimer *self)
+{
+}
+
+static void
+on_aggregate_timer_proxy_created_cb (GObject      *source_object,
+                                     GAsyncResult *result,
+                                     gpointer      user_data)
+{
+  EmtrAggregateTimer *self = user_data;
+  g_autofree gchar *timer_object_path = NULL;
+  g_autoptr(GError) error = NULL;
+
+  self->timer_proxy =
+    emer_aggregate_timer_proxy_new_for_bus_finish (result, &error);
+
+  if (error)
+    g_warning ("Error creating aggregate timer: %s", error->message);
+}
+
+static void
+on_server_aggregate_timer_started_cb (GObject      *source_object,
+                                      GAsyncResult *result,
+                                      gpointer      user_data)
+{
+  EmtrAggregateTimer *self = user_data;
+  g_autofree gchar *timer_object_path = NULL;
+  g_autoptr(GError) error = NULL;
+
+  emer_event_recorder_server_call_start_aggregate_timer_finish (self->dbus_proxy,
```

Comment:
You can pass:

```suggestion
  emer_event_recorder_server_call_start_aggregate_timer_finish (EMER_EVENT_RECORDER_SERVER (source_object),
```

and avoid needing to store `dbus_proxy` in priv.